sympy/matrices/expressions/blockmatrix.py

```diff
@@ -236,6 +236,260 @@ def transpose(self):
         """
         return self._eval_transpose()
 
+    def schur(self, mat = 'A', generalized = False):
+        """Return the Schur Complement of the 2x2 BlockMatrix
+
+        Parameters
+        ==========
+
+        mat : String, optional
+            The matrix with respect to which the
+            Schur Complement is calculated. 'A' is
+            used by default
+
+        generalized : bool, optional
+            If True, returns the generalized Schur
+            Component which uses Moore-Penrose Inverse
+
+        Examples
+        ========
+
+        >>> from sympy import symbols, MatrixSymbol, BlockMatrix
+        >>> m, n = symbols('m n')
+        >>> A = MatrixSymbol('A', n, n)
+        >>> B = MatrixSymbol('B', n, m)
+        >>> C = MatrixSymbol('C', m, n)
+        >>> D = MatrixSymbol('D', m, m)
+        >>> X = BlockMatrix([[A, B], [C, D]])
+
+        The default Schur Complement is evaluated with "A"
+
+        >>> X.schur()
+        -C*A**(-1)*B + D
+        >>> X.schur('D')
+        A - B*D**(-1)*C
```

Comment:
Passing a string like this seems like an odd interface. Would it be possible for the `schur` function to check whether the matrices are invertible and choose which form to use based on that? Otherwise I think that there could be a better way to specify which matrix to invert over.

Reply:
What if there are multiple matrices which are invertible? Maybe we could use the position of the block matrix, i.e. (0,0) for 'A', (0,1) for 'B ' and so on

Reply:
@oscarbenjamin Any thoughts?

Reply:
@oscarbenjamin Please review this when you get some time

Reply:
@oscarbenjamin Please look into this too

Reply:
Right now, although it looks an overkill for use block matrix only to get 2x2 schur complements, I believe that we can make any other place to comprehend schur complements as some result of noncommutative matrix computation later, if there is.